kafka/retry_lazy_client.go

```diff
@@ -0,0 +1,206 @@
+package kafka
+
+import (
+	"time"
+
+	"github.com/Shopify/sarama"
+)
+
+type RetryLazyClient struct {
+	inner  *LazyClient
+	Config *Config
+}
+
+func (c *RetryLazyClient) CreateTopic(t Topic) error {
+	return c.retryOnError(func() error {
+		return c.inner.CreateTopic(t)
+	})
+}
+
+func (c *RetryLazyClient) ReadTopic(name string, refresh_metadata bool) (Topic, error) {
+	return c.retryTopicOnError(func() (Topic, error) {
+		return c.inner.ReadTopic(name, refresh_metadata)
+	})
+}
+
+func (c *RetryLazyClient) UpdateTopic(t Topic) error {
+	return c.retryOnError(func() error {
+		return c.inner.UpdateTopic(t)
+	})
+}
+
+func (c *RetryLazyClient) DeleteTopic(t string) error {
+	return c.retryOnError(func() error {
+		return c.inner.DeleteTopic(t)
+	})
+}
+
+func (c *RetryLazyClient) AddPartitions(t Topic) error {
+	return c.retryOnError(func() error {
+		return c.inner.AddPartitions(t)
+	})
+}
+
+func (c *RetryLazyClient) CanAlterReplicationFactor() (bool, error) {
+	return c.retryBoolOnError(func() (bool, error) {
+		return c.inner.CanAlterReplicationFactor()
+	})
+}
+
+func (c *RetryLazyClient) AlterReplicationFactor(t Topic) error {
+	return c.retryOnError(func() error {
+		return c.inner.AlterReplicationFactor(t)
+	})
+}
+
+func (c *RetryLazyClient) IsReplicationFactorUpdating(topic string) (bool, error) {
+	return c.retryBoolOnError(func() (bool, error) {
+		return c.inner.IsReplicationFactorUpdating(topic)
+	})
+}
+
+func (c *RetryLazyClient) CreateACL(s StringlyTypedACL) error {
+	return c.retryOnError(func() error {
+		return c.inner.CreateACL(s)
+	})
+}
+
+func (c *RetryLazyClient) ListACLs() ([]*sarama.ResourceAcls, error) {
+	return c.retryAclOnError(func() ([]*sarama.ResourceAcls, error) {
+		return c.inner.ListACLs()
+	})
+}
+
+func (c *RetryLazyClient) DeleteACL(s StringlyTypedACL) error {
+	return c.retryOnError(func() error {
+		return c.inner.DeleteACL(s)
+	})
+}
+
+func (c *RetryLazyClient) AlterQuota(q Quota) error {
+	return c.retryOnError(func() error {
+		return c.inner.AlterQuota(q)
+	})
+}
+
+func (c *RetryLazyClient) DescribeQuota(entityType string, entityName string) (*Quota, error) {
+	return c.retryQuotaOnError(func() (*Quota, error) {
+		return c.inner.DescribeQuota(entityType, entityName)
+	})
+}
+
+func (c *RetryLazyClient) retryTimeoutSeconds() time.Duration {
+	return time.Duration(c.Config.RetryTimeout) * time.Second
+}
+
+func (c *RetryLazyClient) retryDelaySeconds() time.Duration {
+	return time.Duration(c.Config.RetryDelay) * time.Second
+}
+
+// Retries allow recovery from transient errors such as connection refused during a broker restart.
+func (c *RetryLazyClient) retryOnError(fn func() error) error {
+	var err error
+	start := time.Now()
+	for {
+		err = fn()
+		if err == nil {
+			return err
+		}
+
+		if time.Since(start) > c.retryTimeoutSeconds() {
+			break
+		}
+
+		//lintignore:R018
```

Comment:
tfproviderlint reports:
    R018: prefer resource.Retry() or (resource.StateChangeConf).WaitForState() over time.Sleep()
Documentation for R018:
    https://github.com/bflad/tfproviderlint/blob/main/passes/R018/README.md

resource.Retry() winds up calling StateChangeConf.WaitForStateContext()
    https://github.com/hashicorp/terraform-plugin-sdk/blob/main/helper/resource/wait.go#L67
    https://github.com/hashicorp/terraform-plugin-sdk/blob/main/helper/resource/state.go#L38

resource.Retry() example usage:
    https://github.com/hashicorp/terraform-website/blob/master/content/plugin/sdkv2/resources/retries-and-customizable-timeouts.mdx

resource.Retry() related documentation:
    https://www.terraform.io/plugin/sdkv2/resources/retries-and-customizable-timeouts

Note that the examples don't use Retry for the actual CRUD operation/request. They use Retry while waiting for the object's remote state to change.
I tried to use resource.Retry and resource.RetryContext, but ran into issues. It can be done but complicates the code.

Reply:
hmm... it seems like bad form to put the whole terraform plan/apply thread to sleep for 15 seconds. the Retry() stuff seems to have better support for task cancelation.